README.md
cache-unit-test
===============

Unit test Python library for testing cache systems like varnish

Based on the work of @onyxfish (https://gist.github.com/onyxfish)



- [x] cachetest.py : productive
- [ ] wordpresscachetest.py : not usable yet

````bash
user@localhost> python3.3 exampletest 
````

# Python test launcher
With the unittest python system, you may have a unittest launcher, and many tests on different folders


````
#!/usr/bin/env python

import sys, getopt
import json
import re
import random
import unittest
import argparse
import cacheunittest
from cacheunittest import CacheUnitTest
import unittest
import os
import sys


if __name__ == "__main__":

	os.environ['MOZ_HEADLESS'] = '1'

	argparse=argparse.ArgumentParser(description="Proxy directives tests app", add_help=True)
	argparse.add_argument("--host", default="127.0.0.1", help="define a target varnish host")
	argparse.add_argument("--httpport", type=int, default=80, help="override the default http port 80")
	argparse.add_argument("--httpsport", type=int, default=443, help="override the default https port 443")
	argparse.add_argument("--verbosity", type=int, default=2, help="verbosity level")
	argparse.add_argument("-p", "--path", default=".", help="code path")
	args = argparse.parse_args()

	verbosity = args.verbosity
	codePath = args.path

	cacheunittest.proxyhost = args.host
	cacheunittest.proxyhttpport  = args.httpport
	cacheunittest.proxyhttpsport = args.httpsport



	print("")
	print("")
	print("Discovering ", codePath, " for artetest*.py tests")
	print("")
	tests = unittest.TestSuite()
	loader = unittest.TestLoader()
	tests_auto = loader.discover(codePath, pattern="artetest*.py", top_level_dir=None)
	result = unittest.TextTestRunner(verbosity=verbosity).run(tests_auto)
	print(result)
	if result.wasSuccessful():
		print("")
		print("")
		print("")
		print("")
		print("    All tests were successfully passed")
		print("        GGGGGOOOOOOOODDDDDD JOB !!!!")
		print("")
		print("")
		print("")
		print("")
		sys.exit(0)
	else:
		print("")
		print("    OOps something wrong happened")
		print("")
		print("###########################")
		print("###########################")
		print("###########################")
		print("#     KEEP THE FAITH!     #")
		print("###########################")
		print("###########################")
		print("###########################")
		sys.exit(2)

````

# cache unit test


# Varnish configuration

The tests are only valid if the varnish configuration contains the following setup
Based on:
* https://www.varnish-software.com/wiki/content/tutorials/varnish/sample_vclTemplate.html
* https://docs.varnish-software.com/tutorials/hit-miss-logging/
* https://www.varnish-software.com/wiki/content/tutorials/varnish/multiple_varnishes.html
* https://varnish-cache.org/docs/trunk/users-guide/purging.html

**It's probably a good idea to put some security around it via a test on client.ip for example**

## Example VCL ([docs/example.vcl](docs/example.vcl)):
````
acl dev {
    # ACL we'll use later to allow purges
    "localhost";
    "127.0.0.1";
    "::1";
}

sub vcl_recv  {

	# To avoid any hack based on these headers
	unset req.http.x-cache;
    unset req.http.X-Back;

    if (req.url ~ "^/static/") {
        set req.http.X-Back = "STATIC";
    } else {
        set req.http.X-Back = "WORDPRESS";
    }

    if (req.method == "PURGE") {
        if (!client.ip ~ dev) {
            return(synth(405,"Not allowed."));
        }
        return (purge);
    }
}

sub vcl_hit {
	set req.http.x-cache = "HIT";
	if (obj.ttl <= 0 and obj.grace > 0s) {
		set req.http.x-cache = "HIT_GRACED";
	}
}

sub vcl_miss {
	set req.http.x-cache = "MISS";
}

sub vcl_pass {
	set req.http.x-cache = "PASS";
}

sub vcl_pipe {
	set req.http.x-cache = "PIPE";
}

sub vcl_synth {
    if (!client.ip ~ dev) {
        set resp.http.X-Back = "VARNISH";
        set resp.http.x-cache = "SYNTH";
    }
}

sub vcl_deliver {
    if (!client.ip ~ dev) {
        if (obj.hits > 0) {
            set resp.http.X-NumberOfHits = obj.hits;
        }
        set resp.http.X-Varnish-Server = server.identity;
        set resp.http.X-Back = req.http.X-Back;
        set resp.http.X-Cache = req.http.X-Cache;
    }
}
````

CHANGES.txt
v1.3.0 -- Ability to play with the content
v1.2.5 -- Remove a bug on the request procedure
v1.2.5 -- Remove debug prints
v1.2.3 -- Corrections
v1.2.0 -- Add HTTPS connection, add any kind of HTTP METHOD request
v1.1.0 -- Add Purge JSON and XML tests
v0.1.12 -- bug fix
v0.1.0 -- Initial release



cachetest.py
import random
import unittest
import sys, getopt
import http.cookiejar
import json
import re
import time

import requests
import http.client

__unittest = True

class CacheTest(unittest.TestCase):
    """
        Basis class common to all CMS and tests
    """


    def setUp(self, proxy, port, website):

        self.website = ""
        self.headers = None

        self.proxies = None
        if proxy != "":
            # print("proxying website ", website, " with proxy ", proxy)
            self.proxies = { "http": proxy, "https": proxy}
            self.proxy = proxy

        self.port = 80
        if port != "":
            self.port = port

        if website != "":
            self.website = website
            self.headers = {'Host' : self.website}

    # Add specific header
    def set_header(self, header_value):
        self.headers.update(header_value)

    def get_request(self, url):
        # Recoding get request to allow proxy
        conn = http.client.HTTPConnection(self.proxy, self.port)
        conn.putrequest("GET", url,  skip_host=True)
        for header in self.headers.keys():
            conn.putheader(str(header), str(self.headers.get(header)))
        conn.endheaders()
        response = conn.getresponse()
        conn.close()
        return response

    def build_url(self, path):
        """
        Construct an absolute url by appending a path to a domain.
        """
        return 'http://%s%s' % (self.website, path)

    def get_once(self, url, **kwargs):
        response = self.get_request(url)
        return response

    def get_twice(self, url, **kwargs):
        """
        Fetch a url twice and return the second response (for testing cache hits).
        """

        self.get_request(url)
        time.sleep(2)
        response = self.get_request(url)
        return response


    def get_twice_tokenized(self, url, tokenname=None, **kwargs):
        """
        Fetch a url twice with two different tokens and return the 2nd response
        """

        if tokenname != None:
            token = tokenname + "=" + str(random.randint(10000,999999))
        else:
            token = str(random.randint(10000,999999))
        #print("url1: " + url + "?" + token)
        self.get_request(url + "?" + token)
        time.sleep(2)
        if tokenname != None    :
            token = tokenname + "=" + str(random.randint(10000,999999))
        else:
            token = str(random.randint(10000,999999))
        #print("url2: " + url + "?" + token)
        response = self.get_request(url + "?" + token)
        return response

    """
        Assertions
    """

    def assertHit(self, response):
        """
        Assert that a given response contains the header indicating a cache hit.
        """
        self.assertEqual(response.headers['X-Cache'].lower(), 'HIT'.lower(), msg='Uncached while cache was expected')

    def assertMiss(self, response):
        """
        Assert that a given response contains the header indicating a cache miss.
        """
        self.assertEqual(response.headers['X-Cache'].lower(), 'miss'.lower())

    def assertMaxAge(self, response, value):
        """
        Assert that a given response contains the header indicating specific "max-age" value.
        """
        MAX_AGE_REGEX = re.compile('max-age\s*=\s*(\d+)')
        try:
            cache_control = response.headers['cache-control']
        except KeyError:
            try:
                cache_control = response.headers['Cache-Control']
            except:
                raise AssertionError('No cache-control header.')

        max_age = MAX_AGE_REGEX.match(cache_control)

        if not max_age:
            raise AssertionError('No max-age specified in cache-control header.')

        self.assertEqual(int(max_age.group(1)), value)

    def assert200(self, response):
        # Ok
        self.assertEqual(response.status, 200)

    def assert30X(self, response):
        self.assertRegex(str(response.status), '30?')

    def assert301(self, response):
        # Permanent redirect
        self.assertEqual(response.status, 301)
 
    def assert302(self, response):
        # Temporary redirect
        self.assertEqual(response.status, 302)
 
    def assert304(self, response):
        # Not modified
        self.assertEqual(response.status, 304)

    def assert40X(self, response):
        self.assertRegex(str(response.status), '40?')

    def assert400(self, response):
        # Bad Request
        self.assertEqual(response.status, 400)

    def assert401(self, response):
        # Unauthorized
        self.assertEqual(response.status, 401)

    def assert403(self, response):
        # Forbidden
        self.assertEqual(response.status, 403)

    def assert404(self, response):
        # Not found
        self.assertEqual(response.status, 404)

    def assert405(self, response):
        # Method Not allowed
        self.assertEqual(response.status, 405)

    def assert50X(self, response):
        # Method Not allowed
        self.assertRegex(str(response.status), '50?')

    def assertBackend(self, response, backend):
        self.assertEqual(response.headers['X-Back'].lower(), backend.lower())

    def assertRedirectURL(self, response, url):
        self.assertEqual(response.reason.lower(), url.lower())




cachetestbase.py
import random
import unittest
import sys, getopt
import http.cookiejar
import json
import re
import time

import requests
import http.client



class CacheTestBase(unittest.TestCase):
    """
        Basis class common to all CMS and tests
    """

    def setUp(self, proxy, port, website):

        self.website = ""
        self.headers = None

        if website != "":
            self.website = website
            self.headers = {'Host' : self.website}


        self.proxies = None
        if proxy != "":
            # print("proxying website ", website, " with proxy ", proxy)
            self.proxies = { "http": proxy, "https": proxy}
            self.proxy = proxy

        self.port = port


    def get_request(self, url):
        # Recoding get request to allow proxy
        conn = http.client.HTTPConnection(self.proxy, self.port)
        conn.putrequest("GET", url,  skip_host=True)
        conn.putheader("Host", self.website)
        conn.endheaders()
        response = conn.getresponse()
        conn.close()
        return response

    def build_url(self, path):
        """
        Construct an absolute url by appending a path to a domain.
        """
        return 'http://%s%s' % (self.website, path)

    def get_once(self, url, **kwargs):
        # self.build_url(url)
        # print("once url ", url, " headers ", self.headers)
        response = self.get_request(url)
        return response

    def get_twice(self, url, **kwargs):
        """
        Fetch a url twice and return the second response (for testing cache hits).
        """
        # self.build_url(url)
        #self.headers.update({'Connection' : 'close'})

        self.get_request(url)
        time.sleep(2)
        # print("Url : ", url, "Request : ", self.headers)
        response = self.get_request(url)
        return response

    """
        Assertions
    """

    def assertHit(self, response):
        """
        Assert that a given response contains the header indicating a cache hit.
        """
        self.assertEqual(response.headers['X-Cache'].lower(), 'HIT'.lower())

    def assertMiss(self, response):
        """
        Assert that a given response contains the header indicating a cache miss.
        """
        self.assertEqual(response.headers['X-Cache'].lower(), 'miss'.lower())

    def assertMaxAge(self, response, value):
        """
        Assert that a given response contains the header indicating specific "max-age" value.
        """
        MAX_AGE_REGEX = re.compile('max-age\s*=\s*(\d+)')
        try:
            cache_control = response.headers['cache-control']
        except KeyError:
            try:
                cache_control = response.headers['Cache-Control']
            except:
                raise AssertionError('No cache-control header.')

        max_age = MAX_AGE_REGEX.match(cache_control)

        if not max_age:
            raise AssertionError('No max-age specified in cache-control header.')

        self.assertEqual(int(max_age.group(1)), value)

    def assert200(self, response):
        # Ok
        self.assertEqual(response.status, 200)

    def assert30X(self, response):
        self.assertRegex(str(response.status), '30?')

    def assert301(self, response):
        # Permanent redirect
        self.assertEqual(response.status, 301)
 
    def assert302(self, response):
        # Temporary redirect
        self.assertEqual(response.status, 302)
 
    def assert304(self, response):
        # Not modified
        self.assertEqual(response.status, 304)

    def assert40X(self, response):
        self.assertRegex(str(response.status), '40?')

    def assert400(self, response):
        # Bad Request
        self.assertEqual(response.status, 400)

    def assert401(self, response):
        # Unauthorized
        self.assertEqual(response.status, 401)

    def assert403(self, response):
        # Forbidden
        self.assertEqual(response.status, 403)

    def assert404(self, response):
        # Not found
        self.assertEqual(response.status, 404)

    def assert405(self, response):
        # Method Not allowed
        self.assertEqual(response.status, 405)

    def assert50X(self, response):
        # Method Not allowed
        self.assertRegex(str(response.status), '50?')

    def assertBackend(self, response, backend):
        self.assertEqual(response.headers['X-Back'].lower(), backend.lower())




cacheunittest.py
import http.client
import http.cookiejar
import json
import random
import re
import ssl
import time
import unittest
from xml.dom import minidom

import logging

logger = logging.getLogger(__name__)

__unittest = True
PROXYHOST = ""
PROXYPORT = ""
PROXYHTTPSPORT = ""

'''
 Based on https://stackoverflow.com/questions/61280350/how-to-set-the-sni-via-http-client-httpsconnection
'''


class WrapSSSLContext(ssl.SSLContext):
	"""
	HTTPSConnection provides no way to specify the
	server_hostname in the underlying socket. We
	accomplish this by wrapping the context to
	overrride the wrap_socket behavior (called later
	by HTTPSConnection) to specify the
	server_hostname that we want.
	"""

	def __new__(cls, server_hostname, *args, **kwargs):
		return super().__new__(cls, *args, *kwargs)

	def __init__(self, server_hostname, *args, **kwargs):
		super().__init__(*args, **kwargs)
		self._server_hostname = server_hostname

	def wrap_socket(self, sock, *args, **kwargs):
		kwargs['server_hostname'] = self._server_hostname
		return super().wrap_socket(sock, *args, **kwargs)


class CacheUnitTest(unittest.TestCase):
	"""
		Basis class common to all CMS and tests
	"""

	def setUp(self, proxy, port, website, https=False):

		self.headers = None

		self.proxies = None
		if proxy:
			self.proxies = {"http": proxy, "https": proxy}
			self.proxy = proxy

		if port:
			self.port = port
		else:
			if not https:
				self.port = 80
			else:
				self.port = 443

		if website:
			self.website = website
			self.headers = {'Host': self.website}

		self.https = https

		self.response_content = ""
		self.response_content_bytes = 2048

	# Add specific header
	def set_header(self, header_value):
		self.headers.update(header_value)

	def purge_request(self, url):
		return self.request("PURGE", url)

	def get_request(self, url):
		return self.request("GET", url)

	def request(self, method, url):
		# Recoding get request to allow proxy
		if not self.https:
			logging.debug("Establishing HTTP Connection with {} on port {}".format(self.proxy, self.port))
			conn = http.client.HTTPConnection(self.proxy, self.port)
		else:
			proxy_to_server_context = WrapSSSLContext(self.proxy)
			logging.debug("Establishing HTTPS Connection with {} on port {}".format(self.proxy, self.port))
			conn = http.client.HTTPSConnection(self.proxy, self.port, context=proxy_to_server_context)

		logging.debug("Pushing request on url {} with method {}".format(url, method))
		conn.putrequest(method, url, skip_host=True)
		for header in self.headers.keys():
			logging.debug("Specifying header {} with value {}".format(str(header), self.headers.get(header)))
			conn.putheader(str(header), str(self.headers.get(header)))
		conn.endheaders()
		response = conn.getresponse()
		self.response_header = response.headers
		self.response_content = str(response.read(self.response_content_bytes))
		conn.close()
		return response

	def build_url(self, path):
		"""
		Construct an absolute url by appending a path to a domain.
		"""
		return 'http://%s%s' % (self.website, path)

	def get_once(self, url, needpurge=False, **kwargs):
		if needpurge:
			self.purge_request(url)
		response = self.get_request(url)
		return response

	def get_twice(self, url, **kwargs):
		"""
		Fetch a url twice and return the second response (for testing cache hits).
		"""

		self.get_request(url)
		# time.sleep(2)
		response = self.get_request(url)
		return response

	def get_twice_tokenized(self, url, tokenname=None, **kwargs):
		"""
		Fetch a url twice with two different tokens and return the 2nd response
		"""

		if tokenname is not None:
			token = tokenname + "=" + str(random.randint(10000, 999999))
		else:
			token = str(random.randint(10000, 999999))
		# print("url1: " + url + "?" + token)
		self.get_request(url + "?" + token)
		if tokenname is not None:
			token = tokenname + "=" + str(random.randint(10000, 999999))
		else:
			token = str(random.randint(10000, 999999))
		# print("url2: " + url + "?" + token)
		response = self.get_request(url + "?" + token)
		return response

	def purgeandget_twice(self, url, **kwargs):
		"""
		Fetch a url twice and return the second response (for testing cache hits).
		"""
		self.purge_request(url)
		time.sleep(1)
		self.get_request(url)
		time.sleep(2)
		response = self.get_request(url)
		return response

	"""
		Assertions
	"""

	def assertHit(self, response):
		"""
		Assert that a given response contains the header indicating a cache hit.
		"""
		self.assertEqual(str(response.headers['X-Cache']).lower(), 'HIT'.lower(),
						 msg='Uncached while cache was expected')

	def assertMiss(self, response):
		"""
		Assert that a given response contains the header indicating a cache miss.
		"""
		self.assertEqual(str(response.headers['X-Cache']).lower(), 'miss'.lower())

	def assertPass(self, response):
		"""
		Assert that a given response contains the header indicating a pass.
		"""
		self.assertEqual(str(response.headers['X-Cache']).lower(), 'pass'.lower())

	def assertSynth(self, response):
		"""
		Assert that a given response contains the header indicating a pass.
		"""
		self.assertEqual(str(response.headers['X-Cache']).lower(), 'synth'.lower())

	def assertMaxAge(self, response, value):
		"""
		Assert that a given response contains the header indicating specific "max-age" value.
		"""
		max_age_regex = re.compile('max-age\s*=\s*(\d+)')
		try:
			cache_control = response.headers['cache-control']
		except KeyError:
			try:
				cache_control = response.headers['Cache-Control']
			except:
				raise AssertionError('No cache-control header.')

		max_age = max_age_regex.match(cache_control)

		if not max_age:
			raise AssertionError('No max-age specified in cache-control header.')

		self.assertEqual(int(max_age.group(1)), value)

	def assert200(self, response):
		# Ok
		self.assertEqual(response.status, 200)

	def assert30X(self, response):
		self.assertRegex(str(response.status), '30?')

	def assert301(self, response):
		# Permanent redirect
		self.assertEqual(response.status, 301)

	def assert302(self, response):
		# Temporary redirect
		self.assertEqual(response.status, 302)

	def assert304(self, response):
		# Not modified
		self.assertEqual(response.status, 304)

	def assert40X(self, response):
		self.assertRegex(str(response.status), '40?')

	def assert400(self, response):
		# Bad Request
		self.assertEqual(response.status, 400)

	def assert401(self, response):
		# Unauthorized
		self.assertEqual(response.status, 401)

	def assert403(self, response):
		# Forbidden
		self.assertEqual(response.status, 403)

	def assert404(self, response):
		# Not found
		self.assertEqual(response.status, 404)

	def assert405(self, response):
		# Method Not allowed
		self.assertEqual(response.status, 405)

	def assert50X(self, response):
		# Method Not allowed
		self.assertRegex(str(response.status), '50?')

	def assertBackend(self, response, backend):
		self.assertEqual(str(response.headers['X-Back']).lower(), backend.lower())

	def assertRedirectURL(self, response, url):
		self.assertEqual(str(response.headers['location']).lower(), url.lower())

	def assertValidJSON(self, response):
		try:
			logging.debug("Parsing response {} first {} bytes, expecting valid JSON".format(self.response_content,
																							self.response_content_bytes))
			json_object = json.loads(self.response_content)
			return True
		except ValueError as error:
			return False

	def assertValidXML(self, response):
		try:
			logging.debug("Parsing response {} first {} bytes, expecting valid JSON".format(self.response_content,
																							self.response_content_bytes))
			minidom.parseString(self.response_content)
			return True
		except ValueError as error:
			return False

